File: app/src/main/java/com/kota/Telnet/TelnetChannel.java
package com.kota.Telnet;

import com.kota.DataPool.MutableByteBuffer;
import com.kota.TextEncoder.B2UEncoder;

import java.io.IOException;
import java.nio.ByteBuffer;

public class TelnetChannel {
    public static int BUFFER_SIZE = 2048;
    private final ByteBuffer _input_buffer = ByteBuffer.allocate(BUFFER_SIZE);
    private TelnetChannelListener _listener = null;
    private boolean _lock = false;
    private final MutableByteBuffer _output_buffer = MutableByteBuffer.createMutableByteBuffer();
    private int _read_data_size = 0;
    private TelnetSocketChannel _socket_channel = null;

    public TelnetChannel(TelnetSocketChannel aSocketChannel) {
        this._socket_channel = aSocketChannel;
    }

    public void setListener(TelnetChannelListener aListener) {
        this._listener = aListener;
    }

    public void clear() {
        this._input_buffer.clear();
        this._output_buffer.clear();
    }

    public void lock() {
        this._lock = true;
    }

    public void unlock() {
        this._lock = false;
    }

    public void writeData(byte[] data) {
        for (byte b : data) {
            writeData(b);
        }
    }

    public void writeData(byte data) {
        this._output_buffer.put(data);
    }

    public boolean sendData() {
        if (this._lock || this._output_buffer.size() <= 0) {
            return false;
        }
        try {
            this._output_buffer.close();
            for (ByteBuffer byteBuffer : this._output_buffer) {
                this._socket_channel.write(byteBuffer);
            }
//            printOutputBuffer();
            this._output_buffer.clear();
            return true;
        } catch (Exception e) {
            e.printStackTrace();
            return false;
        }
    }

    public byte readData() throws TelnetConnectionClosedException, IOException {
        if (!this._input_buffer.hasRemaining()) {
            onReceiveDataStart();
            receiveData();
            onReceiveDataFinished();
        }
        this._input_buffer.mark();
        byte data = this._input_buffer.get();
        this._read_data_size++;
        return data;
    }

    public void undoReadData() {
        this._input_buffer.reset();
        this._read_data_size--;
    }

    private void receiveData() throws TelnetConnectionClosedException, IOException {
        this._input_buffer.clear();
        int read = this._socket_channel.read(this._input_buffer);
        if (this._input_buffer.position() != 0) {
            this._input_buffer.flip();
            return;
        }
        throw new TelnetConnectionClosedException();
    }

    private void onReceiveDataStart() {
        if (this._listener != null) {
            this._listener.onTelnetChannelReceiveDataStart(this);
        }
    }

    private void onReceiveDataFinished() {
        if (this._listener != null) {
            this._listener.onTelnetChannelReceiveDataFinished(this);
//            printInputBuffer();
        }
    }

    /** 列印收到的資料 */
    private void printInputBuffer() {
        byte[] data = new byte[this._input_buffer.limit()];
        for (int i = 0; i < this._input_buffer.limit(); i++) {
            data[i] = this._input_buffer.array()[i];
        }
        try {
            System.out.println("receive data:" + B2UEncoder.getInstance().encodeToString(data));
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    /** 列印送出的資料 */
    private void printOutputBuffer() {
        byte[] data = new byte[this._output_buffer.size()];
        int position = 0;
        for (ByteBuffer output_buffer : this._output_buffer) {
            for (int i = 0; i < output_buffer.limit(); i++) {
                data[position] = output_buffer.array()[i];
                position++;
            }
        }
        try {
            System.out.println("send data:\n" + B2UEncoder.getInstance().encodeToString(data));
            String hex_data = "";
            for (byte datum : data) {
                hex_data = hex_data + String.format(" %1$02x", datum);
            }
            System.out.println("send hex data:\n" + hex_data.substring(1));
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void cleanReadDataSize() {
        this._read_data_size = 0;
    }

    public int getReadDataSize() {
        return this._read_data_size;
    }
}

File: app/src/main/java/com/kota/telnet/TelnetChannel.kt
package com.kota.telnet

import android.util.Log
import com.kota.dataPool.MutableByteBuffer
import com.kota.textEncoder.B2UEncoder
import java.io.IOException
import java.nio.ByteBuffer

class TelnetChannel(aSocketChannel: TelnetSocketChannel) {
    private val inputBuffer: ByteBuffer = ByteBuffer.allocate(BUFFER_SIZE)
    private var channelListener: TelnetChannelListener? = null
    private var isLocked = false
    private val outputBuffer: MutableByteBuffer = MutableByteBuffer.createMutableByteBuffer()
    var readDataSize: Int = 0
        private set
    private var socketChannel: TelnetSocketChannel? = null

    init {
        this.socketChannel = aSocketChannel
    }

    fun setListener(aListener: TelnetChannelListener?) {
        this.channelListener = aListener
    }

    fun clear() {
        this.inputBuffer.clear()
        this.outputBuffer.clear()
    }

    fun lock() {
        this.isLocked = true
    }

    fun unlock() {
        this.isLocked = false
    }

    fun writeData(data: ByteArray) {
        for (b in data) {
            writeData(b)
        }
    }

    fun writeData(data: Byte) {
        this.outputBuffer.put(data)
    }

    fun sendData(): Boolean {
        if (this.isLocked || this.outputBuffer.size() <= 0) {
            return false
        }
        try {
            this.outputBuffer.close()
            for (byteBuffer in this.outputBuffer) {
                this.socketChannel!!.write(byteBuffer)
            }
            //            printOutputBuffer();
            this.outputBuffer.clear()
            return true
        } catch (e: Exception) {
            Log.e(javaClass.simpleName, (if (e.message != null) e.message else "")!!)
            return false
        }
    }

    @Throws(TelnetConnectionClosedException::class, IOException::class)
    fun readData(): Byte {
        if (!this.inputBuffer.hasRemaining()) {
            onReceiveDataStart()
            receiveData()
            onReceiveDataFinished()
        }
        this.inputBuffer.mark()
        val data = this.inputBuffer.get()
        this.readDataSize++
        return data
    }

    fun undoReadData() {
        this.inputBuffer.reset()
        this.readDataSize--
    }

    @Throws(TelnetConnectionClosedException::class, IOException::class)
    private fun receiveData() {
        this.inputBuffer.clear()
        this.socketChannel!!.read(this.inputBuffer)
        if (this.inputBuffer.position() != 0) {
            this.inputBuffer.flip()
            return
        }
        throw TelnetConnectionClosedException
    }

    private fun onReceiveDataStart() {
        if (this.channelListener != null) {
            this.channelListener!!.onTelnetChannelReceiveDataStart(this)
        }
    }

    private fun onReceiveDataFinished() {
        if (this.channelListener != null) {
            this.channelListener!!.onTelnetChannelReceiveDataFinished(this)
            //            printInputBuffer();
        }
    }

    /** 列印收到的資料  */
    private fun printInputBuffer() {
        val data = ByteArray(this.inputBuffer.limit())
        for (i in 0..<this.inputBuffer.limit()) {
            data[i] = this.inputBuffer.array()[i]
        }
        try {
            println("receive data:" + B2UEncoder.instance!!.encodeToString(data))
        } catch (e: Exception) {
            Log.e(javaClass.simpleName, (if (e.message != null) e.message else "")!!)
        }
    }

    /** 列印送出的資料  */
    private fun printOutputBuffer() {
        val data = ByteArray(this.outputBuffer.size())
        var position = 0
        for (outputBuffer in this.outputBuffer) {
            for (i in 0..<outputBuffer!!.limit()) {
                data[position] = outputBuffer.array()[i]
                position++
            }
        }
        try {
            println("send data:\n" + B2UEncoder.instance!!.encodeToString(data))
            var hexData = ""
            for (datum in data) {
                hexData = hexData + String.format(" %1$02x", datum)
            }
            println("send hex data:\n" + hexData.substring(1))
        } catch (e: Exception) {
            Log.e(javaClass.simpleName, (if (e.message != null) e.message else "")!!)
        }
    }

    fun cleanReadDataSize() {
        this.readDataSize = 0
    }

    companion object {
        var BUFFER_SIZE: Int = 2048
    }
}
